docs/docs/09-rfcs/2023-03-08-wing-preview-environment.md

```diff
@@ -0,0 +1,278 @@
+In this case the above comment will contain a direct link for signing up to Wing Cloud.
+After signing up, the developer will be able to create a preview environment for the PR and one will be created automatically after the next code change.
+
+#### Multiple Entry Points
+
+A developer can configure multiple entry points in a single PR.
+Every `*.main.w` file in the repository will be considered as an entry point and will have a dedicated preview environment.
+All preview environments will be available in the PR comment and will be updated upon code changes.
+
+#### Populate Environment with Initial Data
+
+Out of scope for this project.
+
+#### Automatically Running Tests
+
+Upon PR creation and code changes, Wing Cloud Preview Environments will automatically run all tests defined for each entry point.
+For each test run, a new simulator will be created and will be destroyed upon test completion.
+The test results will be available in the PR comment for each entry point.
+
+#### Running Tests Manually
+
+Running tests on your environment is easy using the Wing Console tests UI.
+Each test will reload the simulator to create a clean environment.
+
+Once the test is over it will **not** clean up the simulator data, so you can debug the environment and share outputs with your teammates.
+
+#### User-Controlled Environments termination
+
+Closing a PR will terminate the corresponding preview environments.
+The PR comment will indicate the termination of each preview environment and the urls won't be valid anymore.
+
+#### Self-Cleaning Environments Mechanism
+
+To ensure efficient resource utilization:
+
+1. Preview environments associated with stale PRs (without access or code changes for over 30 days) will automatically deactivate.
+2. The PR's preview environment comment will indicate its inactive status.
+3. Changes in code will trigger redeployment for all related preview environments in this PR.
+
+### Analytics and Logs
+
+To improve the product we are collecting data and monitoring the system.
+
+1. We collect preview environments usage analytics only. We won't collect anything about your code and project.
+2. OpenTelemetry is used for collecting metrics, logs, traces.
+3. Datadog is used for monitoring the system.
+
+### Documentation
+
+Developers reading our docs can easily sign up to Wing Cloud and start using its Preview Environments feature in their repository.
+In our docs we have the following information:
+
+1. What is Wing Cloud Preview Environments
+2. Getting started instructions
+3. How can I configure multiple entry points
+4. Populate environment with initial data
+5. Running tests
+
+### Security
+
+1. The preview environments shouldn’t be able to access files from other preview environments
+2. Only Wing Cloud signed up developers with write access to the repository can create preview environments
+
+## NON Requirements
+
+1. Preview environments will require authentication to view.
```

Comment:
What about non-public github repos? I think if we support them the PE needs to be github authenticated.

Reply:
for the MVP we decided to do what Vercel is doing, which is not to have authentication for PE by default. In order to have auth in Vercel PE you should specifically enable the feature and pay extra for it.
We will support auth protected envs. post MVP

Reply:
So in that case a private vercel repo has a public preview env? If so maybe we need some mangling of the DNS? What does vercel do? Or perhaps just warn the user that since they signed up to wing cloud and the repo is private they should be aware that we're deploying a public PE.

Reply:
We can, although I'll need to design that. There's, again, security implications we have to handle.

Reply:
this is how vercel is doing it:
https://vercel.com/docs/concepts/deployments/deployment-protection#vercel-authentication

TL;DR
- Free to all plans:
enable vercel authentication to all deployed environment. users should be authenticated in vercel before viewing an environment.
- paid
password protected. define specific password for each project.

I think we can start with the first option. WDYT?